### PR TITLE
impr: Speed up getBinaryImages

### PR DESCRIPTION
**Pull request by philipphofmann**
## :scroll: Description

Add two new internal methods `getDebugImagesFromCacheForFrames` and `getDebugImagesFromCacheForThreads` to the `SentryDebugImageProvider`, which use the significantly faster `SentryBinaryImageCache`. https://github.com/getsentry/sentry-cocoa/blob/5a6e387e0b47f0b4fe046839feb957f3fd76c0fc/Sources/Sentry/include/HybridPublic/SentryBinaryImageCache.h#L11-L16

## :bulb: Motivation and Context

Fixes GH-4399

## :green_heart: How did you test it?

Will test it with TestFlight builds.

### Main Branch 
[Event in Sentry](https://sentry-sdks.sentry.io/issues/5485500008/events/b7f50ce02d114e09bef533b88f68240b/?fov=0.7588424437299035%2C75&node=txn-733414a2ea3e42edad6fc450c4c07c2a&project=5428557)

<details><summary>Debug Meta Event JSON</summary>
<p>

```json
{
  "code_file": "/private/var/containers/Bundle/Application/62D5E98F-1727-439B-AEC8-6DD71F97B71D/iOS-Swift.app/iOS-Swift",
  "debug_id": "289a8392-886c-302a-991e-3577820b1c48",
  "arch": "arm64",
  "image_addr": "0x104e58000",
  "image_size": 262144,
  "image_vmaddr": "0x100000000",
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/private/var/containers/Bundle/Application/62D5E98F-1727-439B-AEC8-6DD71F97B71D/iOS-Swift.app/Frameworks/Sentry.framework/Sentry",
  "debug_id": "0889ec9f-e6bb-3a0e-b670-122b9290c1bf",
  "arch": "arm64",
  "image_addr": "0x1052b4000",
  "image_size": 1196032,
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation",
  "debug_id": "4250d45c-51a9-3ac5-8c3d-69de936fad46",
  "arch": "arm64e",
  "image_addr": "0x199251000",
  "image_size": 5529600,
  "image_vmaddr": "0x182d95000",
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices",
  "debug_id": "227122eb-72fd-3ecb-9f66-986addd8e048",
  "arch": "arm64e",
  "image_addr": "0x1e4d39000",
  "image_size": 36864,
  "image_vmaddr": "0x1ce87d000",
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/System/Library/PrivateFrameworks/UIKitCore.framework/UIKitCore",
  "debug_id": "187c890e-8e4c-32f2-a97e-fa905a55f094",
  "arch": "arm64e",
  "image_addr": "0x19ba1c000",
  "image_size": 32247808,
  "image_vmaddr": "0x185560000",
  "debug_status": "found",
  "type": "macho"
}
```

</p>
</details> 

[Transaction in Sentry](https://sentry-sdks.sentry.io/performance/trace/48c2f4b86bdc4eddae7468a06b5b4bf7/?dataset=transactions&field=title&field=project&field=user.display&field=timestamp&field=replayId&id=18214&name=&node=txn-f2f15e9c5b134375862d3ab817371327&project=5428557&query=&queryDataset=transaction-like&sort=-timestamp&source=discover&statsPeriod=1h&timestamp=1728903563&topEvents=5&yAxis=count%28%29)

<details><summary>Transaction Debug Meta JSON</summary>
<p>

```json
{
  "code_file": "/Users/philipp.hofmann/Library/Developer/CoreSimulator/Devices/479CAA0E-3CC9-4B32-BBA2-B6A9E85F1C65/data/Containers/Bundle/Application/CC287D18-AEA8-4E13-B474-BBED1E625104/iOS-Swift.app/iOS-Swift.debug.dylib",
  "debug_id": "30c91ede-2cad-3e55-a6e6-8ef5d437bcde",
  "image_addr": "0x10501c000",
  "image_size": 475136,
  "type": "macho"
}
```

</p>
</details> 

### Feature Branch
[Event in Sentry](https://sentry-sdks.sentry.io/issues/5485405120/events/1bf3c876835d46be87af920aefd3f15a/?fov=1%2C110&node=txn-c44fa8c44d454bb9bd47602747610a09&project=5428557)

<details><summary>Debug Meta Event JSON</summary>
<p>

```json
{
  "code_file": "/private/var/containers/Bundle/Application/90A1CD24-E825-411B-A3CB-9E6796A3EF74/iOS-Swift.app/Frameworks/Sentry.framework/Sentry",
  "debug_id": "0ca81d91-985b-3ede-81e1-a868f53b4cce",
  "arch": "arm64",
  "image_addr": "0x10534c000",
  "image_size": 1196032,
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation",
  "debug_id": "4250d45c-51a9-3ac5-8c3d-69de936fad46",
  "arch": "arm64e",
  "image_addr": "0x199251000",
  "image_size": 5529600,
  "image_vmaddr": "0x182d95000",
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices",
  "debug_id": "227122eb-72fd-3ecb-9f66-986addd8e048",
  "arch": "arm64e",
  "image_addr": "0x1e4d39000",
  "image_size": 36864,
  "image_vmaddr": "0x1ce87d000",
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/private/var/containers/Bundle/Application/90A1CD24-E825-411B-A3CB-9E6796A3EF74/iOS-Swift.app/iOS-Swift",
  "debug_id": "289a8392-886c-302a-991e-3577820b1c48",
  "arch": "arm64",
  "image_addr": "0x104e98000",
  "image_size": 262144,
  "image_vmaddr": "0x100000000",
  "debug_status": "found",
  "type": "macho"
},
{
  "code_file": "/System/Library/PrivateFrameworks/UIKitCore.framework/UIKitCore",
  "debug_id": "187c890e-8e4c-32f2-a97e-fa905a55f094",
  "arch": "arm64e",
  "image_addr": "0x19ba1c000",
  "image_size": 32247808,
  "image_vmaddr": "0x185560000",
  "debug_status": "found",
  "type": "macho"
}
```
</p>
</details> 

[Transaction in Sentry](https://sentry-sdks.sentry.io/performance/trace/4a3f03f74c874eeb81d3dd5327f1a43e/?dataset=transactions&field=title&field=project&field=user.display&field=timestamp&field=replayId&id=18214&name=&node=span-2868becdc23d4e39&node=txn-1aa5ab9c58894c1eaad37a32e89e2a1e&project=5428557&query=&queryDataset=transaction-like&sort=-timestamp&source=discover&statsPeriod=1h&timestamp=1728903256&topEvents=5&yAxis=count%28%29)

<details><summary>Transaction Debug Meta JSON</summary>
<p>

```json
{
  "code_file": "/Users/philipp.hofmann/Library/Developer/CoreSimulator/Devices/479CAA0E-3CC9-4B32-BBA2-B6A9E85F1C65/data/Containers/Bundle/Application/2EA49ABC-0881-420B-9533-E0A962158479/iOS-Swift.app/iOS-Swift.debug.dylib",
  "debug_id": "30c91ede-2cad-3e55-a6e6-8ef5d437bcde",
  "image_addr": "0x1010a0000",
  "image_size": 475136,
  "type": "macho"
}
```

</p>
</details> 

### Profiles

**Main Branch**
CleanShot 2024-10-14 at 13 10 46@2x

**Feature Branch**
CleanShot 2024-10-14 at 13 09 26@2x

The profile of the feature branch doesn't show anything suspicious for NSScanner, which the `sentry_UInt64ForHexAddress` uses. The profile on the main branch highlights that `sentrycrashdl_getBinaryImageForHeader` has 2.00 Mc and weight of 0.4%. The `getDebugImagesForThreads` has a weight of 2,0% on the main branch, but the `getDebugImagesFromCacheForThreads` has only 0.3% weight. This proves that the new implementation of `getDebugImagesFromCacheForThreads` should be significantly faster.

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [x] I updated the docs if needed.
- [x] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: Sources/Sentry/SentryBinaryImageCache.m [hunk 1]
```diff
@@ -80,6 +83,17 @@ - (void)binaryImageAdded:(const SentryCrashBinaryImage *)image
     }
 }
 
++ (NSString *_Nullable)convertUUID:(const unsigned char *const)value
```
Comment (philipphofmann) [inline Sources/Sentry/SentryBinaryImageCache.m:86]: Moved from `SentryDebugImageProvider` to here.

File: .github/workflows/testflight.yml [hunk 2]
```diff
@@ -15,6 +15,8 @@ on:
     paths:
       - '.github/workflows/testflight.yml'
 
+# Only to trigger testflight upload for testing. Remove before merging.
+
```
Comment (brustolin) [inline .github/workflows/testflight.yml:19]: m: Forgot the comment

**Review by brustolin — APPROVED** [review]
LGTM

File: Sources/Sentry/SentryDebugImageProvider.m [hunk 3]
```diff
@@ -95,6 +103,55 @@ - (void)extractDebugImageAddressFromFrames:(NSArray<SentryFrame *> *)frames
...
+    NSMutableSet<NSString *> *imageAddresses = [[NSMutableSet alloc] init];
+
+    for (SentryThread *thread in threads) {
+        [self extractDebugImageAddressFromFrames:thread.stacktrace.frames intoSet:imageAddresses];
+    }
+
+    return [self getDebugImagesForImageAddressesFromCache:imageAddresses];
+}
+
+- (NSArray<SentryDebugMeta *> *)getDebugImagesForImageAddressesFromCache:
```
[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryDebugImageProvider.m:125]: Yes, I can also expose it for Hybrid SDKs. I didn't want to make it public, though.

[the remark this answers is not in the document]
Comment (philipphofmann) [reply Sources/Sentry/SentryDebugImageProvider.m:125]: Here you go https://github.com/getsentry/sentry-cocoa/pull/4435/commits/641c1e1ea3bbe3d670b68ddad9fb096626144e51.

File: .github/workflows/testflight.yml (hunk 2 above)
Comment (philipphofmann) [reply .github/workflows/testflight.yml]: Thanks, I ditched it.

File: Sources/Sentry/SentryDebugImageProvider.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (krystofwoldrich) [reply Sources/Sentry/SentryDebugImageProvider.m:125]: Awesome, thank you.

**Pull request merged**